plugin/src/test/java/io/jenkins/plugins/forensics/git/reference/GitCheckoutListenerMergeITest.java

```diff
@@ -0,0 +1,234 @@
+
+/**
+ * Integration test which verifies the {@link GitCheckoutListener} in case of merges.
+ *
+ * @author Florian Orendi
+ */
+class GitCheckoutListenerMergeITest extends GitITest {
+
+    private static final String MAIN_JOB_NAME = "Main_Job";
+    private static final String PR_JOB_NAME = "PR_Job";
+    private static final String PR_BRANCH_NAME = "pull-request-branch";
+    private static final String FEATURE_FILE = "feature_file";
+    private static final String MAIN_FILE = "new_main_file";
+
+    @Test
+    void shouldHandleMergeFromMainIntoPullRequest() throws IOException {
+        FreeStyleProject mainJob = createFreeStyleProject(
+                MAIN_JOB_NAME, MAIN_JOB_NAME, "origin/" + INITIAL_BRANCH);
+        FreeStyleProject pullRequestJob = createFreeStyleProject(
+                PR_JOB_NAME, MAIN_JOB_NAME, "origin/" + PR_BRANCH_NAME);
+
+        buildSuccessfully(mainJob);
+
+        checkoutNewBranch(PR_BRANCH_NAME);
+        commitFile(FEATURE_FILE, "Commit 1 in PR");
+        buildSuccessfully(pullRequestJob);
+        commitFile(FEATURE_FILE, "Commit 2 in PR");
+        buildSuccessfully(pullRequestJob);
+
+        checkout(INITIAL_BRANCH);
+        commitFile(MAIN_FILE, "Commit 1 in main");
+        buildSuccessfully(mainJob);
+        commitFile(MAIN_FILE, "Commit 2 in main");
+        Run<?, ?> latestMainBuild = buildSuccessfully(mainJob);
+        GitCommitsRecord latestMainRecord = latestMainBuild.getAction(GitCommitsRecord.class);
+        assertThat(latestMainRecord)
+                .isNotNull()
+                .satisfies(commit -> {
+                    assertThat(commit.getTargetParentCommit()).isEqualTo(ObjectId.zeroId().name());
+                    assertThat(commit.getLatestCommit()).isEqualTo(getHead());
+                });
+
+        checkout(PR_BRANCH_NAME);
+        commitFile("anotherFile", "Commit 3 in PR");
+        Run<?, ?> latestPullRequestBuild = buildSuccessfully(pullRequestJob);
+        GitCommitsRecord latestPullRequestRecord = latestPullRequestBuild.getAction(GitCommitsRecord.class);
+        assertThat(latestPullRequestRecord)
+                .isNotNull()
+                .satisfies(commit -> {
+                    assertThat(commit.getTargetParentCommit()).isEqualTo(ObjectId.zeroId().name());
+                    assertThat(commit.getLatestCommit()).isEqualTo(getHead());
+                });
+
+        // merge main into PR and create a merge commit
+        mergeWithoutFastForwarding(INITIAL_BRANCH);
+        // verify commits
+        Run<?, ?> mergeBuild = buildSuccessfully(pullRequestJob);
+        assertThat(mergeBuild.getAction(GitCommitsRecord.class))
+                .isNotNull()
+                .satisfies(commit -> {
+                    assertThat(commit.getLatestCommit()).isEqualTo(getHead());
+                    assertThat(commit.getTargetParentCommit()).isEqualTo(latestMainRecord.getLatestCommit());
+                });
+    }
```

Comment:
You are right.
The reference finder did not observe merge commits or target parent commits since these commits have not been stored respectively investigated before.
Now, since these merge commits are available, they are also used for finding the reference build.

I added some assertions to verify the behavior. The new test class `GitMergeITest` verifies the behavior after merges.